From e0e4ac1bbc056b3489361ce840f5c3de3478725b Mon Sep 17 00:00:00 2001
From: Szehon Ho <szehon@cloudera.com>
Date: Wed, 5 Aug 2015 17:34:10 -0700
Subject: [PATCH 0527/1164] CDH-38095: HIVE-11319 : CTAS with location
 qualifier overwrites directories (Yongzhi Chen
 via Szehon)

Conflicts:
	ql/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java

Change-Id: If56f038479954d820c83c41efaf5d5ee08d18217
---
 .../java/org/apache/hadoop/hive/ql/ErrorMsg.java   |    1 +
 .../hadoop/hive/ql/parse/SemanticAnalyzer.java     |   26 ++++++++++++++++++++
 .../queries/clientnegative/ctas_noemptyfolder.q    |   10 ++++++++
 .../clientnegative/ctas_noemptyfolder.q.out        |   19 ++++++++++++++
 4 files changed, 56 insertions(+)
 create mode 100644 ql/src/test/queries/clientnegative/ctas_noemptyfolder.q
 create mode 100644 ql/src/test/results/clientnegative/ctas_noemptyfolder.q.out

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java b/ql/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java
index 419168b..b9ce385 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/ErrorMsg.java
@@ -429,6 +429,7 @@
       "Alter table partition type {0} does not support cascade", true),
 
   DROP_NATIVE_FUNCTION(10301, "Cannot drop native function"),
+  CTAS_LOCATION_NONEMPTY(10304, "CREATE-TABLE-AS-SELECT cannot create table with location to a non-empty directory."),
   COMPILE_LOCK_TIMED_OUT(10308, "Attempt to acquire compile lock timed out.", true),
   //========================== 20000 range starts here ========================//
   SCRIPT_INIT_ERROR(20000, "Unable to initialize custom script."),
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java b/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
index 241353b..7b42cce 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java
@@ -20,6 +20,7 @@
 
 import static org.apache.hadoop.hive.conf.HiveConf.ConfVars.HIVESTATSDBCLASS;
 
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.Serializable;
 import java.security.AccessControlException;
@@ -11136,6 +11137,31 @@ ASTNode analyzeCreateTable(
         throw new SemanticException(e);
       }
 
+      if(location != null && location.length() != 0) {
+        Path locPath = new Path(location);
+        FileSystem curFs = null;
+        FileStatus locStats = null;
+        try {
+          curFs = locPath.getFileSystem(conf);
+          if(curFs != null) {
+            locStats = curFs.getFileStatus(locPath);
+          }
+          if(locStats != null && locStats.isDir()) {
+            FileStatus[] lStats = curFs.listStatus(locPath);
+            if(lStats != null && lStats.length != 0) {
+              throw new SemanticException(ErrorMsg.CTAS_LOCATION_NONEMPTY.getMsg(location));
+            }
+          }
+        } catch (FileNotFoundException nfe) {
+          //we will create the folder if it does not exist.
+        } catch (IOException ioE) {
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Exception when validate folder ",ioE);
+          }
+
+        }
+      }
+
       tblProps = addDefaultProperties(tblProps);
 
       crtTblDesc = new CreateTableDesc(qualifiedTabName[0], dbDotTab, isExt, isTemporary, cols,
diff --git a/ql/src/test/queries/clientnegative/ctas_noemptyfolder.q b/ql/src/test/queries/clientnegative/ctas_noemptyfolder.q
new file mode 100644
index 0000000..18c9086
--- /dev/null
+++ b/ql/src/test/queries/clientnegative/ctas_noemptyfolder.q
@@ -0,0 +1,10 @@
+create table ctas1 
+location 'file:${system:test.tmp.dir}/ctastmpfolder' 
+as 
+select * from src limit 3;
+
+create table ctas2
+location 'file:${system:test.tmp.dir}/ctastmpfolder'
+as 
+select * from src limit 2;
+
diff --git a/ql/src/test/results/clientnegative/ctas_noemptyfolder.q.out b/ql/src/test/results/clientnegative/ctas_noemptyfolder.q.out
new file mode 100644
index 0000000..76d7b86
--- /dev/null
+++ b/ql/src/test/results/clientnegative/ctas_noemptyfolder.q.out
@@ -0,0 +1,19 @@
+PREHOOK: query: create table ctas1 
+#### A masked pattern was here ####
+as 
+select * from src limit 3
+PREHOOK: type: CREATETABLE_AS_SELECT
+PREHOOK: Input: default@src
+#### A masked pattern was here ####
+PREHOOK: Output: database:default
+PREHOOK: Output: default@ctas1
+POSTHOOK: query: create table ctas1 
+#### A masked pattern was here ####
+as 
+select * from src limit 3
+POSTHOOK: type: CREATETABLE_AS_SELECT
+POSTHOOK: Input: default@src
+#### A masked pattern was here ####
+POSTHOOK: Output: database:default
+POSTHOOK: Output: default@ctas1
+#### A masked pattern was here ####
-- 
1.7.9.5

